src/messaging/ExchangeMessageDispatch.cpp

```diff
@@ -51,45 +51,60 @@ CHIP_ERROR ExchangeMessageDispatch::SendMessage(SessionManager * sessionManager,
     PayloadHeader payloadHeader;
     payloadHeader.SetExchangeID(exchangeId).SetMessageType(protocol, type).SetInitiator(isInitiator);
 
-    // If there is a pending acknowledgment piggyback it on this message.
-    if (reliableMessageContext->HasPiggybackAckPending())
+    if (session->AllowsMRP())
     {
-        payloadHeader.SetAckMessageCounter(reliableMessageContext->TakePendingPeerAckMessageCounter());
-    }
-
-    if (IsReliableTransmissionAllowed() && reliableMessageContext->AutoRequestAck() &&
-        reliableMessageContext->GetReliableMessageMgr() != nullptr && isReliableTransmission)
-    {
-        auto * reliableMessageMgr = reliableMessageContext->GetReliableMessageMgr();
-
-        payloadHeader.SetNeedsAck(true);
-
-        ReliableMessageMgr::RetransTableEntry * entry = nullptr;
-
-        // Add to Table for subsequent sending
-        ReturnErrorOnFailure(reliableMessageMgr->AddToRetransTable(reliableMessageContext, &entry));
-        auto deleter = [reliableMessageMgr](ReliableMessageMgr::RetransTableEntry * e) {
-            reliableMessageMgr->ClearRetransTable(*e);
-        };
-        std::unique_ptr<ReliableMessageMgr::RetransTableEntry, decltype(deleter)> entryOwner(entry, deleter);
-
-        ReturnErrorOnFailure(sessionManager->PrepareMessage(session, payloadHeader, std::move(message), entryOwner->retainedBuf));
-        CHIP_ERROR err = sessionManager->SendPreparedMessage(session, entryOwner->retainedBuf);
-        err            = ReliableMessageMgr::MapSendError(err, exchangeId, isInitiator);
-        ReturnErrorOnFailure(err);
-        reliableMessageMgr->StartRetransmision(entryOwner.release());
+        // If there is a pending acknowledgment piggyback it on this message.
+        if (reliableMessageContext->HasPiggybackAckPending())
+        {
+            payloadHeader.SetAckMessageCounter(reliableMessageContext->TakePendingPeerAckMessageCounter());
+        }
+
+        if (IsReliableTransmissionAllowed() && reliableMessageContext->AutoRequestAck() &&
+            reliableMessageContext->GetReliableMessageMgr() != nullptr && isReliableTransmission)
```

Comment:
This is also pre-existing, but if we get here and `isReliableTransmission`, then `reliableMessageContext->AutoRequestAck()` is in fact true.  And `AllowsMRP()` is true.

So this seems like it would be much simpler if the `AllowsMRP()` check were just around the piggyback ack thing.  And if you really want, inside this block where isReliableTransmission is true you could VerifyOrDie(AllowsMRP()) or some equivalent, because that boolean must never be true if !AllowsMRP().

And then you would not need any other changes to this code at all.

Reply:
It turns out IsReliableTransmissionAllowed() is hardcoded to true. Might be the same as AllowsMRP() FWIW and makes it more confusing. 
But my point in this change was to make connection-oriented flows bypass all these logic and not wade through them at all. So, I have encapsulated all these MRP specific changes within AllowsMRP() which TCP and BLE should bypass. That, I think, keeps the code cleaner and simpler I guess, no? Ultimately, code-change size wise, it might turn out to be almost similar.

Reply:
> It turns out IsReliableTransmissionAllowed() is hardcoded to true

It's a virtual method.  Some test bits subclass it and return false.